### PR TITLE
Add function to transpose from a local vector to distributed one and vice versa

### PR DESCRIPTION
**Pull request by KFilipek**
This PR introduces two functions to transpose, where one argument is a local data structure.

### CODE REVIEW

**Comment by rscohn2** [comment]
CTest logs are saved as artifacts. The fails are:
```
[  FAILED  ] CpuMpiTests.local_to_distribute_transpose
```

**Comment by KFilipek** [comment]
@rscohn2 @lslusarczyk this PR is ready to review

File: test/gtest/cpu/mdspan.cpp [hunk 1]
```diff
@@ -167,3 +169,58 @@ TEST(CpuMpiTests, DISABLED_distributed_mdarray_transpose) {
...
+    lib::collective::transpose(0, std::make_optional(lsrc), ddst);
+  } else {
+    lib::collective::transpose(
+        0, std::optional<std::experimental::mdspan<T, dyn_2d>>(), ddst);
+  }
+
+  if (comm_rank == 0) {
+    EXPECT_EQ(lsrc(0, 0), v[0]);
+    EXPECT_EQ(lsrc(1, 0), ddst(0, 1));
+    EXPECT_EQ(lsrc(rows - 1, cols - 2), ddst(cols - 2, rows - 1));
```
Comment (lslusarczyk) [inline test/gtest/cpu/mdspan.cpp:224]: Can we try to use `check_eq` method here? It checks all items in the matrix. Currently it is executed on local mdspans only but I guess it should run well on distributed too.

**Review by lslusarczyk — CHANGES_REQUESTED** [review]
A few comments to @KFilipek.
I'd like to try a one small refactor related to `optionals` here, so @rscohn2 please hold with review for a while.

File: test/gtest/cpu/mdspan.cpp (hunk 1 above)
Comment (KFilipek) [reply test/gtest/cpu/mdspan.cpp]: I've added this change as a separate commit

**Comment by rscohn2** [comment]
> I'd like to try a one small refactor related to `optionals` here, so @rscohn2 please hold with review for a while.

Not sure if you are done. It looks ok to me. The optional seems better than nullptr when it fits. I am thinking that we should not support nullptr as a direct argument. User can assign nullptr to an appropriately typed pointer that is passed. It would simplify the types and be more natural SPMD style for me.

**Comment by rscohn2** [comment]
In the near term, we just need something that works for transpose and continuing with `mdspan` is ok.

**Comment by lslusarczyk** [comment]
@rscohn2 , please take a look at https://github.com/oneapi-src/distributed-ranges/pull/24 PR, which consists of this changes + remove of optional, as mdspan can be empty itself. If it looks good to you let's close this PR and merge that PR.

**Comment by rscohn2** [comment]
Superseded by #24

**Pull request closed without merging**